File: multi-app/src/MultiApp.Batch/Jobs/OrderExpiryJob.cs
using MultiApp.Domain.Enums;
using MultiApp.Domain.Interfaces;
using MultiApp.Domain.Services;

namespace MultiApp.Batch.Jobs;

public class OrderExpiryJob
{
    private readonly IOrderRepository _repository;
    private readonly OrderDomainService _domainService;
    private readonly TimeSpan _expiryWindow;

    public OrderExpiryJob(IOrderRepository repository, OrderDomainService domainService, TimeSpan expiryWindow)
    {
        _repository = repository;
        _domainService = domainService;
        _expiryWindow = expiryWindow;
    }

    public int Run()
    {
        var pending = _repository.GetByStatus(OrderStatus.Pending);
        var expired = _domainService.FindExpired(pending, _expiryWindow);

        foreach (var order in expired)
        {
            order.Expire();
            _repository.Save(order);
        }

        return expired.Count;
    }
}

File: java-multi-app/batch/src/main/java/com/example/batch/OrderExpiryJob.java
package com.example.batch;

import com.example.core.model.Order;
import com.example.infrastructure.InMemoryOrderRepository;
import java.util.List;

public class OrderExpiryJob {
    private final InMemoryOrderRepository repository;

    public OrderExpiryJob(InMemoryOrderRepository repository) {
        this.repository = repository;
    }

    public int run() {
        List<Order> all = repository.findAll();
        int expired = 0;
        for (Order order : all) {
            if ("PENDING".equals(order.getStatus())) {
                order.setStatus("EXPIRED");
                repository.save(order);
                expired++;
            }
        }
        return expired;
    }
}
